File: backend/mno-integration/adapters/mtn.py
import logging
import requests
from typing import Dict, Any
from tenacity import retry, stop_after_attempt, wait_fixed, retry_if_exception_type, before_sleep_log

from .base import MNOAdapter
from .fallback import SFTPFallbackManager

logger = logging.getLogger(__name__)

# This mock exception simulates what would be thrown by the Tenacity CircuitBreaker
# when the 3-failure threshold inside 30 seconds is reached.
class CircuitBreakerOpenError(Exception):
    pass

class MTNAdapter(MNOAdapter):
    """
    Adapter for MTN Nigeria's mobile data deduction REST API.
    Implements a strict circuit breaker and fallback SFTP queue.
    """
    
    def __init__(self, api_base_url: str):
        super().__init__("MTN")
        self.api_base_url = api_base_url
        self.fallback_manager = SFTPFallbackManager("MTN")

    # Retry logic: Try up to 3 times with 1-second waits on ConnectionError or Timeout
    @retry(
        stop=stop_after_attempt(3),
        wait=wait_fixed(1),
        retry=retry_if_exception_type((requests.exceptions.ConnectionError, requests.exceptions.Timeout)),
        before_sleep=before_sleep_log(logger, logging.WARNING),
        reraise=True
    )
    def _execute_rest_call(self, imsi: str, amount_mb: float) -> requests.Response:
        """Isolated REST call to allow Tenacity to cleanly trap retries/failures."""
        payload = {
            "imsi": imsi,
            "deduct_amount": amount_mb,
            "unit": "MB"
        }
        
        # Simulate an actual POST to MTN's charging gateway
        # In this mock, we force a connect error to demonstrate the breaker
        raise requests.exceptions.ConnectionError("MTN API Gateway unreachable")
        
        # response = requests.post(f"{self.api_base_url}/v1/charging/deduct", json=payload, timeout=2.0)
        # response.raise_for_status()
        # return response

    def deduct_balance(self, imsi: str, amount_mb: float) -> Dict[str, Any]:
        try:
            # Attempt to hit the primary REST API
            response = self._execute_rest_call(imsi, amount_mb)
            
            # If successful, parse result
            data = response.json()
            return {
                "status": "SUCCESS",
                "remaining_balance": data.get("balance"),
                "source": "REST_API"
            }
            
        except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
            # Circuit breaker flipped open (simulated by Tenacity giving up after 3 tries)
            logger.error(f"[MTN] Circuit Breaker OPEN after 3 failures. Engaging SFTP Fallback.")
            
            return self.fallback_sftp_batch(imsi, amount_mb)
        except Exception as e:
            logger.error(f"[MTN] Unhandled exception during deduction: {e}")
            return {"status": "ERROR", "message": str(e)}

    def fallback_sftp_batch(self, imsi: str, amount_mb: float) -> Dict[str, Any]:
        """
        Engages the fallback mechanism, storing the deduction for later batch processing.
        """
        success = self.fallback_manager.append_record(imsi, amount_mb)
        if success:
            return {
                "status": "QUEUED_SFTP",
                "message": "Deduction logged to offline batch file for later sync.",
                "source": "SFTP_FALLBACK"
            }
        return {"status": "ERROR", "message": "SFTP Fallback also failed."}

File: backend/mno-integration/adapters/mtn.go
package adapters

import (
	"errors"
	"log"
	"time"

	"github.com/sony/gobreaker"
)

type MTNAdapter struct {
	apiBaseUrl      string
	fallbackManager *SFTPFallbackManager
	cb              *gobreaker.CircuitBreaker
}

func NewMTNAdapter(apiBaseUrl string) *MTNAdapter {
	cbSettings := gobreaker.Settings{
		Name:        "MTN",
		MaxRequests: 1,
		Interval:    30 * time.Second,
		Timeout:     60 * time.Second,
		ReadyToTrip: func(counts gobreaker.Counts) bool {
			// Trip circuit breaker after 3 consecutive failures
			return counts.ConsecutiveFailures >= 3
		},
		OnStateChange: func(name string, from gobreaker.State, to gobreaker.State) {
			log.Printf("[MTN] Circuit Breaker state changed from %v to %v", from, to)
		},
	}

	return &MTNAdapter{
		apiBaseUrl:      apiBaseUrl,
		fallbackManager: NewSFTPFallbackManager("MTN", ""),
		cb:              gobreaker.NewCircuitBreaker(cbSettings),
	}
}

func (a *MTNAdapter) CarrierName() string {
	return "MTN"
}

func (a *MTNAdapter) executeRestCall(imsi string, amountMB float64) (map[string]interface{}, error) {
	// Simulate an actual POST to MTN's charging gateway
	// In this mock, we force a connect error to demonstrate the breaker
	time.Sleep(50 * time.Millisecond)
	return nil, errors.New("MTN API Gateway unreachable")
}

func (a *MTNAdapter) DeductBalance(imsi string, amountMB float64) DeductionResult {
	res, err := a.cb.Execute(func() (interface{}, error) {
		return a.executeRestCall(imsi, amountMB)
	})

	if err != nil {
		if errors.Is(err, gobreaker.ErrOpenState) || errors.Is(err, gobreaker.ErrTooManyRequests) || err.Error() == "MTN API Gateway unreachable" {
			log.Printf("[MTN] Circuit Breaker OPEN or request failed. Engaging SFTP Fallback.")
			return a.FallbackSFTPBatch(imsi, amountMB)
		}
		
		log.Printf("[MTN] Unhandled exception during deduction: %v", err)
		return DeductionResult{
			Status:  "ERROR",
			Message: err.Error(),
		}
	}

	data := res.(map[string]interface{})
	balance, _ := data["balance"].(float64)

	return DeductionResult{
		Status:           "SUCCESS",
		RemainingBalance: balance,
		Source:           "REST_API",
	}
}

func (a *MTNAdapter) FallbackSFTPBatch(imsi string, amountMB float64) DeductionResult {
	success := a.fallbackManager.AppendRecord(imsi, amountMB)
	if success {
		return DeductionResult{
			Status:  "QUEUED_SFTP",
			Message: "Deduction logged to offline batch file for later sync.",
			Source:  "SFTP_FALLBACK",
		}
	}
	return DeductionResult{
		Status:  "ERROR",
		Message: "SFTP Fallback also failed.",
	}
}
